pkg/controllers/routesync/reconcile.go

```diff
@@ -0,0 +1,220 @@
+)
+
+var (
+	httpRouteTarget = routev1.RouteTargetReference{
+		Kind:   "Service",
+		Name:   httpServiceName,
+		Weight: pointer.Int32Ptr(100),
+	}
+	httpsRouteTarget = routev1.RouteTargetReference{
+		Kind:   "Service",
+		Name:   httpsServiceName,
+		Weight: pointer.Int32Ptr(100),
+	}
+)
+
+type processRoute string
+
+const (
+	skipRoute   processRoute = "skip"
+	updateRoute processRoute = "update"
+	createRoute processRoute = "create"
+)
+
+// RouteSyncReconciler holds the fields necessary to run the Route Sync reconciliation
+type RouteSyncReconciler struct {
+	TargetClient routeclient.Interface
+	HostClient   routeclient.Interface
+	Namespace    string // Note: the target cluster name and the namespace it resides in are the same
+	TargetLister routelister.RouteLister
+	HostLister   routelister.RouteLister
+	Log          logr.Logger
+}
+
+// Reconcile will react to any Route change and reconcile all the remote target routes
+// against the local host list of routes. It will create local host routes and point them
+// to the appropriate proxy service for delivery into the target cluster's routers.
+func (r *RouteSyncReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	r.Log.Info("Reconciling target routes")
+
+	targetRoutes, err := r.TargetLister.List(labels.Everything())
+	if err != nil {
+		r.Log.Error(err, "failed to list target cluster routes")
+		return ctrl.Result{}, err
+	}
+
+	currentHostRoutes, err := r.HostLister.List(labels.Everything())
+	if err != nil {
+		r.Log.Error(err, "failed to list host routes")
+		return ctrl.Result{}, err
+	}
+
+	routesToCreate := []*routev1.Route{}
+	routesToUpdate := []*routev1.Route{}
+	routesToSkip := []*routev1.Route{}
+	routesToDelete := []*routev1.Route{}
+
+	// Walk through each remote target route, and identify which needs
+	// creation, update, or skipping.
+	for _, targetRoute := range targetRoutes {
+		syncRoute, action := r.createSyncRouteFromTarget(currentHostRoutes, targetRoute)
+
+		switch action {
+		case createRoute:
+			routesToCreate = append(routesToCreate, syncRoute)
+		case updateRoute:
+			routesToUpdate = append(routesToUpdate, syncRoute)
+		case skipRoute:
+			routesToSkip = append(routesToSkip, syncRoute)
+		}
+	}
+
+	allProposedHostRoutes := listOfAllRoutes(routesToCreate, routesToUpdate, routesToSkip)
+	// Walk through each local host route and identify any that need to be removed.
+	for i, hostRoute := range currentHostRoutes {
+		if hostRoute.Labels[routeLabel] != r.Namespace {
+			// ignore routes that this controller isn't responsible for
+			continue
+		}
+		if hostRouteNeedsDeletion(allProposedHostRoutes, hostRoute) {
+			routesToDelete = append(routesToDelete, currentHostRoutes[i])
+		}
+	}
+
+	for _, route := range routesToCreate {
+		r.Log.Info("Will create route", "route", route.Name)
+		_, err := r.HostClient.RouteV1().Routes(r.Namespace).Create(route)
+		if err != nil {
+			r.Log.Error(err, "failed to create route")
```

Comment:
It may be better to keep a running list of errors and then report a single aggregated error at the end, that way a problem with a single route doesn't block the rest of the updates.

Reply:
Same with the other loops below